### PR TITLE
Fix incorrect cast in StrictOptimizedMapOps.+

### LINKED ISSUE #12699: Unsafe cast in StrictOptimisedMapOps.+ in 2.13
## Reproduction steps

Scala version: 2.13.x

```scala

object MapTest {
  def main(args: Array[String]): Unit = {
    val m1: collection.immutable.HashMap[Int, Int] = HashMap(1 -> 1);
    // class scala.collection.immutable.$colon$colon cannot be cast to class scala.collection.immutable.Map (scala.collection.immutable.$colon$colon and scala.collection.immutable.Map are in unnamed module of loader 'app')
    //	at demo.MapTest$.main(Test.scala:68)
    //	at demo.MapTest.main(Test.scala)
    //
    // This is s "delayed" CCE from the asInstanceOf that is compiled to a no-op in StrictOptimizedMapOps.+
    //
    //   @deprecated("Use ++ with an explicit collection argument instead of + with varargs", "2.13.0")
    //  override def + [V1 >: V](elem1: (K, V1), elem2: (K, V1), elems: (K, V1)*): CC[K, V1] = {
    //    val m = ((this + elem1).asInstanceOf[Map[K, V]] + elem2).asInstanceOf[CC[K, V1]]
    //    if(elems.isEmpty) m else m.concat(elems).asInstanceOf[CC[K, V1]]
    //
    val m2: Map[Int, Any] = m1.+(elem1 = 4 -> 4, elem2 = 5 -> 5, elems = List(6 -> 6, 7 -> 7, 8 -> 8, 9 -> 9): _*)
    val m3: Map[Int, Any] = m1 + (4 -> 4, 5 -> 5, 6 -> 6, 7 -> 7, 8 -> 8, 9 -> 9)
  }
}
```

## Problem

Need to use `m.mapFactory.newBuilder` to get the suitable return type.

### PR DESCRIPTION
**Pull request by lrytz**
Fixes https://github.com/scala/bug/issues/12699

### CODE REVIEW

**Review by som-snytt — APPROVED** [review]
Tests pass locally. The best part is adding a space after `if`.

**Comment by som-snytt** [comment]
Oh, not spurious. I couldn't see the cause
```
[error] 1 of 19 test tasks failed:
[error] - junit/testOnly -- +v
[error]   - junit/test:compileIncremental failed: Compilation failed
[error] Failure due to previous errors
[error] (testAll) Failure due to previous errors
```
but travis shows
```
[error] /home/travis/build/scala/scala/test/junit/scala/collection/immutable/MapTest.scala:149:25: multiarg infix syntax looks like a tuple and will be deprecated

[error]     assertEquals(7, (m1 + (2 -> 2, 3 -> 3, 4 -> 4, 5 -> 5, 6 -> 6, 7 -> 7)).size)

[error]                         ^

[error] one error found
```
Ha, ha.

Fixable by adding `@deprecated` to test method.

Could also remove extra space `List( 4 -> 4,` while editing.
Could also remove parens from test method, as that does not warn in test.

**Comment by som-snytt** [comment]
Travis also reports `downloading sbt launcher 1.6.2`. Not sure at what point that matters. TIL `testOnly -- +v`.

**Comment by som-snytt** [comment]
Edit: I tried deprecating the test method and it still errors out locally. At least TIL `set Global / fatalWarnings := true` which FSR isn't `Werror`.

Oh I didn't read the message carefully:
```
context.warning(tree.pos, "multiarg infix syntax looks like a tuple and will be deprecated", WarningCategory.LintMultiargInfix)
```
This was a pre-deprecation warning. I'll try `@predeprecated`.

Also this makes my blood boil every time:
```
MapTest.scala:145:4: not found: type nowarn
```
Of course it's also necessary to suppress the deprecation of the method.

File: test/junit/scala/collection/immutable/MapTest.scala [hunk 1]
```diff
@@ -141,4 +141,11 @@ class MapTest {
       }
     }
   }
+
```
Comment (som-snytt) [inline test/junit/scala/collection/immutable/MapTest.scala:146]:
```suggestion
  @deprecated("tests deprecated API", "")
  @nowarn("cat=lint-multiarg-infix")
```

**Comment by som-snytt** [comment]
Probably lrytz didn't notice that I added a commit...

I assumed he was asleep, I didn't check the time zones.

I must have done it wrong: https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/16757/
```
+import annotation._
+
 import org.junit.Assert.assertEquals
 import org.junit.Test

@@ -142,10 +144,12 @@ class MapTest {
     }
   }

+  @deprecated("tests deprecated API", since="2.13.11")
+  @nowarn("cat=lint-multiarg-infix")
   @Test
-  def t12699(): Unit = {
+  def t12699: Unit = {
     val m1: HashMap[Int, Int] = HashMap(1 -> 1)
-    assertEquals(7, m1.+(elem1 = 2 -> 2, elem2 = 3 -> 3, elems = List( 4 -> 4, 5 -> 5, 6 -> 6, 7 -> 7): _*).size)
+    assertEquals(7, m1.+(elem1 = 2 -> 2, elem2 = 3 -> 3, elems = List(4 -> 4, 5 -> 5, 6 -> 6, 7 -> 7): _*).size)
     assertEquals(7, (m1 + (2 -> 2, 3 -> 3, 4 -> 4, 5 -> 5, 6 -> 6, 7 -> 7)).size)
   }
 }
```

**Comment by lrytz** [comment]
I'm sorry :-/ I saw the fatal warning three hours ago and now just came back to fix it without paying attention. Feel free to push -f e876cb7 again!

**Comment by som-snytt** [comment]
I don't understand how it passes with your nowarn, so I won't attempt it again. Later, I'll try to figure it out, maybe after I do some advent of code.

**Comment by lrytz** [comment]
I'm using `m1.+` instead of infix `m1 +`, the nowarn is for `+` being deprecated...

**Comment by som-snytt** [comment]
Thanks, my eyes were too bleary to see it.

**Pull request merged**